File: 101.symmetric-tree.py
#
# @lc app=leetcode id=101 lang=python3
#
# [101] Symmetric Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.check_symmetry(root.left, root.right) if root else root

    def check_symmetry(self, left, right):
        if not left or not right:
            return left == right
        if left.val != right.val:
            return False
        return self.check_symmetry(left.left, right.right) and self.check_symmetry(
            left.right, right.left
        )


# @lc code=end

File: 101.symmetric-tree.rs
/*
 * @lc app=leetcode id=101 lang=rust
 *
 * [101] Symmetric Tree
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {

    }

    pub fn check_symmetry(left: Option<Rc<RefCell<TreeNode>>>, right: Option<Rc<RefCell<TreeNode>>>) -> bool {

        if None == left || None == right {
            return true;
        }
        match (left, right) {
            (None, None) => true,
            (Some(l), Some(r)) => {
                let (l, r) = (l.borrow(), r.borrow());
                l.val == r.val && Self::check_symmetry(l.left.clone(), r.right.clone()) && Self::check_symmetry(l.right.clone(), r.left.clone())
            },
            _ => false
        }
    }
}
// @lc code=end

